lib/trino-hive-formats/src/test/java/io/trino/hive/formats/ion/TestIonFormat.java

```diff
@@ -0,0 +1,242 @@
+                List.of(new SqlDecimal(BigInteger.valueOf(123400), 10, 2), new SqlDecimal(BigInteger.valueOf(123400000), 25, 5)),
+                List.of(new SqlDecimal(BigInteger.valueOf(120000), 10, 2), new SqlDecimal(BigInteger.valueOf(120000000), 25, 5)),
+                List.of(new SqlDecimal(BigInteger.valueOf(123400), 10, 2), new SqlDecimal(BigInteger.valueOf(123400000), 25, 5)));
+    }
+
+    @Test
+    public void testOversizeOrOverpreciseDecimals()
+    {
+        // todo: implement
+    }
+
+    @Test
+    public void testEncode()
+            throws IOException
+    {
+        List<Column> columns = List.of(
+                new Column("magic_num", INTEGER, 0),
+                new Column("some_text", VARCHAR, 1),
+                new Column("is_summer", BooleanType.BOOLEAN, 2),
+                new Column("byte_clob", VarbinaryType.VARBINARY, 3),
+                new Column("sequencer", new ArrayType(INTEGER), 4),
+                new Column("struction", RowType.rowType(
+                        field("foo", INTEGER),
+                        field("bar", VARCHAR)), 5));
+
+        List<?> row1 = List.of(17, "hello", true, new SqlVarbinary(new byte[] {(byte) 0xff}), List.of(1, 2, 3), List.of(51, "baz"));
+        List<?> row2 = List.of(31, "goodbye", false, new SqlVarbinary(new byte[] {(byte) 0x01, (byte) 0xaa}), List.of(7, 8, 9), List.of(67, "qux"));
+
+        Page page = toPage(columns, row1, row2);
+
+        writeAndPrint(columns, page);
+    }
+
+    private void assertValues(RowType rowType, String ionText, List<Object>... expected)
+            throws IOException
+    {
+        final List<RowType.Field> fields = rowType.getFields();
+        final List<Column> columns = IntStream.range(0, fields.size())
+                .boxed()
+                .map(i -> {
+                    final RowType.Field field = fields.get(i);
+                    return new Column(field.getName().get(), field.getType(), i);
+                })
+                .toList();
+        final IonDecoder decoder = IonDecoderFactory.buildDecoder(columns);
+        final PageBuilder pageBuilder = new PageBuilder(expected.length, rowType.getFields().stream().map(RowType.Field::getType).toList());
+
+        try (IonReader ionReader = IonReaderBuilder.standard().build(ionText)) {
+            for (int i = 0; i < expected.length; i++) {
+                assertThat(ionReader.next()).isNotNull();
+                pageBuilder.declarePosition();
+                decoder.decode(ionReader, pageBuilder);
+            }
+            assertThat(ionReader.next()).isNull();
+        }
+
+        for (int i = 0; i < expected.length; i++) {
+            final List<Object> actual = readTrinoValues(columns, pageBuilder.build(), i);
+            assertColumnValuesEquals(columns, actual, expected[i]);
+        }
+    }
+
+    // todo: this needs to assert, not print to stderr!
```

Comment:
Is something blocking this? Might be nice for the test coverage.

Reply:
Nothing other than that I didn't get to it. It is _covered_ by the overall tests in the plugin submodule, but having assertions here would help for targeting the actual writing of data.

Reply:
I added assertions based on IonValue equivalence.